luavm/src/lib/lua_values/mod.rs

```diff
@@ -0,0 +1,754 @@
+        }
+    }
+
+    pub fn add(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? + other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? + other.to_int()?)
+        })
+    }
+
+    pub fn sub(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? - other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? - other.to_int()?)
+        })
+    }
+
+    pub fn mul(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? * other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? * other.to_int()?)
+        })
+    }
+
+    pub fn div(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? / other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? / other.to_int()?)
+        })
+    }
+
+    pub fn modulus(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from(self.to_float()? % other.to_float()?)
+        } else {
+            LuaVal::from(self.to_int()? % other.to_int()?)
+        })
+    }
+
+    pub fn fdiv(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(if self.is_float() || other.is_float() {
+            LuaVal::from((self.to_float()? / other.to_float()?).floor())
+        } else {
+            LuaVal::from(self.to_int()? / other.to_int()?)
+        })
+    }
+
+    pub fn exp(&self, other: &LuaVal) -> Result<LuaVal, LuaError> {
+        Ok(LuaVal::from(self.to_float()?.powf(other.to_float()?)))
+    }
+}
+
+impl From<i64> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(int: i64) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(int) ^ LuaValKind::INT as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<f64> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(float: f64) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(float) ^ LuaValKind::FLOAT as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<String> for LuaVal {
+    /// Create a string LuaVal.
+    fn from(string: String) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(string) ^ LuaValKind::STR as usize,
+            owns: true,
+        }
+    }
+}
+
+impl From<LuaTable> for LuaVal {
+    /// Create an integer LuaVal.
+    fn from(table: LuaTable) -> Self {
+        LuaVal {
+            encoded_ptr: to_raw_ptr(Gc::new(table)) ^ LuaValKind::TABLE as usize,
+            owns: true,
+        }
+    }
+}
+
+impl Drop for LuaVal {
+    fn drop(&mut self) {
+        if self.owns {
+            unsafe {
```

Comment:
The `unsafe` is covering too much code here, and needs to be moved inside the relevant bits of code.